diff --git a/drivers/media/i2c/ov5640.c b/drivers/media/i2c/ov5640.c
index 42869b40745b..36d91ad10a6b 100644
--- a/drivers/media/i2c/ov5640.c
+++ b/drivers/media/i2c/ov5640.c
@@ -134,7 +134,8 @@ enum ov5640_mode_id {
 };
 
 enum ov5640_frame_rate {
-	OV5640_15_FPS = 0,
+	OV5640_5_FPS = 0,
+	OV5640_15_FPS,
 	OV5640_30_FPS,
 	OV5640_60_FPS,
 	OV5640_NUM_FRAMERATES,
@@ -344,6 +345,7 @@ MODULE_PARM_DESC(virtual_channel,
 		 "MIPI CSI-2 virtual channel (0..3), default 0");
 
 static const int ov5640_framerates[] = {
+	[OV5640_5_FPS] = 5,
 	[OV5640_15_FPS] = 15,
 	[OV5640_30_FPS] = 30,
 	[OV5640_60_FPS] = 60,
@@ -1018,9 +1020,9 @@ static const struct ov5640_mode_info ov5640_mode_data[OV5640_NUM_MODES] = {
 		.csi2_timings = {
 			.analog_crop = {
 				.left	= 0,
-				.top	= 250,
+				.top	= 0,
 				.width	= 2624,
-				.height	= 1456,
+				.height	= 1952,
 			},
 			.crop = {
 				.left	= 16,
@@ -1028,18 +1030,18 @@ static const struct ov5640_mode_info ov5640_mode_data[OV5640_NUM_MODES] = {
 				.width	= 1280,
 				.height	= 720,
 			},
-			.htot		= 1600,
-			.vblank_def	= 560,
+			.htot		= 2844,
+			.vblank_def	= 24,
 		},
 		.reg_data	= ov5640_setting_720P_1280_720,
 		.reg_data_size	= ARRAY_SIZE(ov5640_setting_720P_1280_720),
 		.max_fps	= OV5640_30_FPS,
-		.def_fps	= OV5640_30_FPS
+		.def_fps	= OV5640_15_FPS
 	}, {
 		/* 1920x1080 */
 		.id		= OV5640_MODE_1080P_1920_1080,
 		.dn_mode	= SCALING,
-		.pixel_rate	= OV5640_PIXEL_RATE_148M,
+		.pixel_rate	= OV5640_PIXEL_RATE_124M,
 		.width		= 1920,
 		.height		= 1080,
 		.dvp_timings = {
@@ -1073,18 +1075,18 @@ static const struct ov5640_mode_info ov5640_mode_data[OV5640_NUM_MODES] = {
 				.width	= 1920,
 				.height	= 1080,
 			},
-			.htot		= 2234,
-			.vblank_def	= 24,
+			.htot		= 2500,
+			.vblank_def	= 40,
 		},
 		.reg_data	= ov5640_setting_1080P_1920_1080,
 		.reg_data_size	= ARRAY_SIZE(ov5640_setting_1080P_1920_1080),
-		.max_fps	= OV5640_30_FPS,
-		.def_fps	= OV5640_30_FPS
+		.max_fps	= OV5640_15_FPS,
+		.def_fps	= OV5640_5_FPS
 	}, {
 		/* 2592x1944 */
 		.id		= OV5640_MODE_QSXGA_2592_1944,
 		.dn_mode	= SCALING,
-		.pixel_rate	= OV5640_PIXEL_RATE_168M,
+		.pixel_rate	= OV5640_PIXEL_RATE_124M,
 		.width		= OV5640_PIXEL_ARRAY_WIDTH,
 		.height		= OV5640_PIXEL_ARRAY_HEIGHT,
 		.dvp_timings = {
@@ -1122,8 +1124,8 @@ static const struct ov5640_mode_info ov5640_mode_data[OV5640_NUM_MODES] = {
 		},
 		.reg_data	= ov5640_setting_QSXGA_2592_1944,
 		.reg_data_size	= ARRAY_SIZE(ov5640_setting_QSXGA_2592_1944),
-		.max_fps	= OV5640_15_FPS,
-		.def_fps	= OV5640_15_FPS
+		.max_fps	= OV5640_5_FPS,
+		.def_fps	= OV5640_5_FPS
 	},
 };
 
@@ -2743,11 +2745,11 @@ static int ov5640_try_frame_interval(struct ov5640_dev *sensor,
 				     u32 width, u32 height)
 {
 	const struct ov5640_mode_info *mode;
-	enum ov5640_frame_rate rate = OV5640_15_FPS;
+	enum ov5640_frame_rate rate = OV5640_5_FPS;
 	int minfps, maxfps, best_fps, fps;
 	int i;
 
-	minfps = ov5640_framerates[OV5640_15_FPS];
+	minfps = ov5640_framerates[OV5640_5_FPS];
 	maxfps = ov5640_framerates[OV5640_60_FPS];
 
 	if (fi->numerator == 0) {
@@ -2916,9 +2918,7 @@ static int ov5640_update_pixel_rate(struct ov5640_dev *sensor)
 	__v4l2_ctrl_modify_range(sensor->ctrls.hblank,
 				 hblank, hblank, 1, hblank);
 
-	vblank = timings->vblank_def;
-
-	if (sensor->current_fr != mode->def_fps) {
+	{
 		/*
 		 * Compute the vertical blanking according to the framerate
 		 * configured with s_frame_interval.
@@ -3646,8 +3646,7 @@ static int ov5640_s_frame_interval(struct v4l2_subdev *sd,
 	}
 
 	if (ov5640_framerates[frame_rate] > ov5640_framerates[mode->max_fps]) {
-		ret = -EINVAL;
-		goto out;
+		frame_rate = mode->max_fps;
 	}
 
 	if (mode != sensor->current_mode ||
